Documentation for a PowerDNS development environment based on a language server (#11377)
PowerDNS/pdns: DEVELOPMENT.md

On some systems, `clangd` and `clang-tidy` are available as packages separate from the `clang` package.
Ensure that you have all three binaries available and running on your system.

# Compilation Database

For projects with non-trivial build systems, like PowerDNS, `clangd` requires a [compilation database](https://clang.llvm.org/docs/JSONCompilationDatabase.html).

Since PowerDNS' autotools-based build system does not have native support for generating such a database, an external tool like [Bear (the Build EAR)](https://github.com/rizsotto/Bear) can be used.

Once you have `bear` installed, configure a build of your choice (either the PowerDNS `auth`, `recursor` or `dnsdist`) using `clang` and `clang++`:

```sh
make distclean    # Ensure we rebuild all files so that bear can pick them up.
CC=clang CXX=clang++ ./configure --with-modules=gsqlite3 --disable-lua-records
```

We can now build PowerDNS using `bear` and `make` which produces a compilation database file named `compile_commands.json`:

```sh
bear --append -- make -j 8
```

Once the compilation database is generated, you can now move onto setting up an LSP client in your editor or IDE.

> Note that the process of generating the compilation database file only needs to be run only when a file is added to the project or when build flags change (e.g. dependencies are added).

Reviewer: ```suggestion
Note that the process of generating the compilation database file only needs to be run when a file is added to the project or when build flags change (e.g. dependencies are added).
```

Code snippets below should be added to your Emacs init file (e.g. `~/.config/emacs/init`).

We'll start by enabling Emacs package repositories and declaring which packages we would like to have installed:

```elisp
(with-eval-after-load 'package
 (setq package-archives
  '(("gnu"   . "https://elpa.gnu.org/packages/")
    ("melpa" . "https://melpa.org/packages/")))
 (push 'company      package-selected-packages)
 (push 'flycheck     package-selected-packages)
 (push 'lsp-mode     package-selected-packages)
 (push 'lsp-ui       package-selected-packages)
 (push 'lsp-treemacs package-selected-packages))
```

To avoid restarting Emacs, you can evaluate that previous s-expression by pointing at the last parenthesis and using `C-x C-e` or selecting the block and using `M-x eval-region`.

Once done, run `M-x package-refresh-contents`, then `M-x package-install-selected-packages`.
This should install some packages for you.

Now let's set up our common programming mode, this enables the following features:

> * Highlighting of the current line.
> * Display of line numbers.
> * Auto-insertion of indentation.
> * Auto-insertion of closing parenthesis, brackets, etc...

Reviewer: ```suggestion
* Highlighting the current line.
* Displaying line numbers.
* Auto-inserting indentation.
* Auto-inserting closing parenthesis, brackets, etc...
```

* Auto-completion.
* On-the-fly code checking.

> * On-the-fly spell checking for code.

Reviewer: ```suggestion
* On-the-fly spell checking code.
```

> * Highlighting of matching parenthesis, brackets, etc...
> * Auto-display of documentation briefs in the echo area.

Reviewer: ```suggestion
* Highlighting matching parentheses, brackets, etc...
* Auto-displaying documentation briefs in the echo area.
```

 (setq lsp-ui-doc-use-webkit t))

(with-eval-after-load 'lsp-ui-peek
 (setq lsp-ui-peek-list-width 30)
 (setq lsp-ui-peek-always-show t))

(with-eval-after-load 'lsp-ui-sideline
 (setq lsp-ui-sideline-enable nil))

(with-eval-after-load 'lsp-ui
 (define-key lsp-ui-mode-map (kbd "M-."    ) #'lsp-ui-peek-find-definitions)
 (define-key lsp-ui-mode-map (kbd "M-?"    ) #'lsp-ui-peek-find-references)
 (define-key lsp-ui-mode-map (kbd "M-I"    ) #'lsp-ui-peek-find-implementation)
 (define-key lsp-ui-mode-map (kbd "C-c d"  ) #'lsp-ui-doc-show)
 (define-key lsp-ui-mode-map (kbd "C-c ! l") #'lsp-ui-flycheck-list))
```

And finally, set up the C/C++ programming mode with a few settings:

* Indentation of 2 spaces.
* Simple auto-detection of coding style.
* Marking of badly-styled comments.
* Running the LSP client.

```elisp
(defmacro setup-c-style-comments ()
 "Setup C-style /* ... */ comments."
 `(with-eval-after-load 'newcomment
   (setq-local comment-style 'extra-line)))

(with-eval-after-load 'cc-mode
 (add-hook 'c-mode-common-hook #'lsp))

(with-eval-after-load 'cc-vars
 (setq c-mark-wrong-style-of-comment t)
 (setq c-default-style '((other . "user")))
 (setq c-basic-offset 2)

>  (add-hook 'c-mode-common-hook (lambda nil (progn (setup-c-style-comments)))))

Reviewer: I suspect this should be `set-up-c...`

PowerDNS Development Environment
--------------------------------

Thank you for you interest to contribute to the PowerDNS project. This

> document will explain one way to setup a development environment based

Reviewer: ```suggestion
document will explain one way to set up a development environment based
```

on the language server protocol (LSP) when working on PowerDNS.

# Introduction

The environment will consist of setting up the
[`clangd`](https://clangd.llvm.org/) C/C++ language server to enable
fancy IDE features for
development. [`ccls`](https://github.com/MaskRay/ccls) can also be used
in place of `clangd`.

> Furthermore, additional on-the-fly checks using `clang-tidy` can be

Reviewer: ```suggestion
Furthermore, additional [on-the-fly checks using `clang-tidy` ](#on-the-fly-clang-tidy) can be
```

> easily enabled [as shown below](DEVELOPMENT.md#on-the-fly-clang-tidy).

Reviewer: ```suggestion
easily enabled.
```

This is another _don't say click here_.

On some systems, `clangd` and `clang-tidy` are available as packages

> that are separate from the `clang` package. Ensure that you have all

Reviewer: ```suggestion
separate from the `clang` package. Ensure that you have all
```

Mode](https://emacs-lsp.github.io/) for C/C++ development using `clangd`
which supports [on-the-fly
checking](https://www.flycheck.org/en/latest/) and
[auto-completion](https://company-mode.github.io/), among many other
features.

Code snippets below should be added to your Emacs init file
(e.g. `~/.config/emacs/init`).

We'll start by enabling Emacs package repositories and declaring which
packages we would like to have installed:

```elisp
(with-eval-after-load 'package
 (setq package-archives
  '(("gnu"   . "https://elpa.gnu.org/packages/")
    ("melpa" . "https://melpa.org/packages/")))
 (push 'company      package-selected-packages)
 (push 'flycheck     package-selected-packages)
 (push 'lsp-mode     package-selected-packages)
 (push 'lsp-ui       package-selected-packages)
 (push 'lsp-treemacs package-selected-packages))
```

To avoid restarting Emacs, you can evaluate that previous s-expression
by pointing at the last parenthesis and using `C-x C-e` or selecting the
block and using `M-x eval-region`.

Once done, run `M-x package-refresh-contents`, then `M-x
package-install-selected-packages`. This should install some packages for you.

> Now let's setup our common programming mode, this enables the following

Reviewer: ```suggestion
Now let's set up our common programming mode, this enables the following
```

I'm not promising to flag each instance.

(n) setup
(vp) set up

# Editors

## Emacs

This section explains how to set up [Emacs](https://www.gnu.org/software/emacs/) with [LSP Mode](https://emacs-lsp.github.io/) for C/C++ development using `clangd` which supports [on-the-fly checking](https://www.flycheck.org/en/latest/) and [auto-completion](https://company-mode.github.io/), among many other features.

Instructions for an alternative, more minimal, setup using [Eglot](https://github.com/joaotavora/eglot) [are also available](#minimal-emacs).

Code snippets below should be added to your Emacs init file (e.g. `~/.config/emacs/init`).

We'll start by enabling Emacs package repositories and declaring which packages we would like to have installed:

```elisp
(with-eval-after-load 'package
 (setq package-archives
  '(("gnu"   . "https://elpa.gnu.org/packages/")
    ("melpa" . "https://melpa.org/packages/")))
 (push 'company      package-selected-packages)
 (push 'flycheck     package-selected-packages)
 (push 'lsp-mode     package-selected-packages)
 (push 'lsp-ui       package-selected-packages)
 (push 'lsp-treemacs package-selected-packages))
```

> To avoid restarting Emacs, you can evaluate that previous s-expression by pointing at the last parenthesis and using `C-x C-e` or selecting the block and using `M-x eval-region`.

Reviewer: This is correct. You're asking a user to focus a single `)`.

Once done, run `M-x package-refresh-contents`, then `M-x package-install-selected-packages`.
This should install some packages for you.

Now let's set up our common programming mode, this enables the following features:

* Highlighting the current line.
* Displaying line numbers.
* Auto-inserting indentation.

> * Auto-inserting closing parenthesis, brackets, etc...

Reviewer: This can't be right as is.

```suggestion
* Auto-inserting closing parentheses, brackets, etc...
```

This would be fine, it's inserting in the mass sense plural `)` and plural `]`s.

Reviewer: You could possibly do this instead:
```suggestion
* Auto-inserting closing parenthesis, bracket, etc...
```

Inserting a single `)` or a single `]`.

 (push 'lsp-treemacs package-selected-packages))
```

> To avoid restarting Emacs, you can evaluate that previous s-expression by pointing at the last parentheses and using `C-x C-e` or selecting the block and using `M-x eval-region`.

Reviewer: This is actually different, it's really pointing at the last (singular) `parenthesis`.

Unlike the item below which refers to inserting (plural) `parentheses` and `brackets`.
